### PR TITLE
feat: save external controller info during consume

### PR DESCRIPTION
**Pull request by SimonRichardson**
This is part 1 of a series of change sets that implement the consume application offer.

----

When consuming an application offer, we need to save the external controller information. This is only required if it's cross controller relations.

This was already implemented, but ripped out when we had to remove all the state calls. Thus, it was easy to put back, with a lot more tests.

----

This pull request implements the backend logic for consuming remote applications (cross-model relations, or CMR) in the Juju application facade. It introduces the ability to save external controller information when consuming an application offer, updates the API and its tests to support this, and adds comprehensive test coverage for the new functionality.

Key changes include:

**CMR Backend Implementation:**

* Implements the `Consume` method in `APIBase`, enabling the addition of remote applications to a model and saving external controller details if provided. This includes new logic for parsing and validating source model and controller tags, and updating external controller info only when necessary.
* Adds the `ExternalControllerService` to the `APIBase` struct and wiring, allowing the backend to update external controller records as part of the consume process. [[1]](diffhunk://#diff-c547d7b2b5fbb9bb1e7d8f6b3d9d14f811d713eafdec1908f91f90a52110cb4aR91) [[2]](diffhunk://#diff-c547d7b2b5fbb9bb1e7d8f6b3d9d14f811d713eafdec1908f91f90a52110cb4aR163) [[3]](diffhunk://#diff-c547d7b2b5fbb9bb1e7d8f6b3d9d14f811d713eafdec1908f91f90a52110cb4aR241) [[4]](diffhunk://#diff-f1c79d0a3fbec4c0c70fd4dd06085bdf56c351f679b0cb5d69b6234237f20cacR36) [[5]](diffhunk://#diff-f1c79d0a3fbec4c0c70fd4dd06085bdf56c351f679b0cb5d69b6234237f20cacR69) [[6]](diffhunk://#diff-f1c79d0a3fbec4c0c70fd4dd06085bdf56c351f679b0cb5d69b6234237f20cacR140)

**API and Struct Enhancements:**

* Adds `controllerUUID` to `APIBase` and ensures it is passed and set during construction, so the backend can distinguish between local and external controllers. [[1]](diffhunk://#diff-c547d7b2b5fbb9bb1e7d8f6b3d9d14f811d713eafdec1908f91f90a52110cb4aR78) [[2]](diffhunk://#diff-c547d7b2b5fbb9bb1e7d8f6b3d9d14f811d713eafdec1908f91f90a52110cb4aR177) [[3]](diffhunk://#diff-c547d7b2b5fbb9bb1e7d8f6b3d9d14f811d713eafdec1908f91f90a52110cb4aR208) [[4]](diffhunk://#diff-c547d7b2b5fbb9bb1e7d8f6b3d9d14f811d713eafdec1908f91f90a52110cb4aR232) [[5]](diffhunk://#diff-f1c79d0a3fbec4c0c70fd4dd06085bdf56c351f679b0cb5d69b6234237f20cacR56) [[6]](diffhunk://#diff-f1c79d0a3fbec4c0c70fd4dd06085bdf56c351f679b0cb5d69b6234237f20cacR91) [[7]](diffhunk://#diff-f1c79d0a3fbec4c0c70fd4dd06085bdf56c351f679b0cb5d69b6234237f20cacR154)
* Updates imports and type signatures to include new dependencies (`crossmodel`, `uuid`, etc.) and support the new functionality. [[1]](diffhunk://#diff-c547d7b2b5fbb9bb1e7d8f6b3d9d14f811d713eafdec1908f91f90a52110cb4aR29-R30) [[2]](diffhunk://#diff-602c8325d9b9625d6bec625289b29eb023a3e1806b91e60920e58fb952534717R20) [[3]](diffhunk://#diff-9818e721d50e51e2503652dec1c4229b2a4f2881e97e684478ec8084ba01c3ceR18-R24) [[4]](diffhunk://#diff-9818e721d50e51e2503652dec1c4229b2a4f2881e97e684478ec8084ba01c3ceR46-R62) [[5]](diffhunk://#diff-f1c79d0a3fbec4c0c70fd4dd06085bdf56c351f679b0cb5d69b6234237f20cacR20-R23) [[6]](diffhunk://#diff-f1c79d0a3fbec4c0c70fd4dd06085bdf56c351f679b0cb5d69b6234237f20cacR69)

**Testing Improvements:**

* Adds a comprehensive suite of tests for the new `Consume` method, covering permission checks, argument validation, external controller updates, and various edge cases.
* Updates test setup and mocks to support the new `ExternalControllerService` and `controllerUUID` requirements. [[1]](diffhunk://#diff-881d7a76b221e0e7665b4e659043c6a4d8081fdebd1df4326d6e575fab3dec79L30-R37) [[2]](diffhunk://#diff-881d7a76b221e0e7665b4e659043c6a4d8081fdebd1df4326d6e575fab3dec79L39-R53) [[3]](diffhunk://#diff-881d7a76b221e0e7665b4e659043c6a4d8081fdebd1df4326d6e575fab3dec79L591-R605) [[4]](diffhunk://#diff-881d7a76b221e0e7665b4e659043c6a4d8081fdebd1df4326d6e575fab3dec79L668-R682) [[5]](diffhunk://#diff-f1c79d0a3fbec4c0c70fd4dd06085bdf56c351f679b0cb5d69b6234237f20cacR140)
* Adjusts test generator directives to include `ExternalControllerService` in mocks.

These changes lay the groundwork for full cross-model relation support by enabling backend handling of remote application consumption and external controller management.

## QA steps

```sh
$ juju bootstrap lxd src
$ juju add-model source
$ juju deploy juju-qa-test
$ juju offer juju-qa-test:juju-info
$ juju offers
```

```sh
$ juju bootstrap lxd dst
$ juju add-model dest
```

```sh
$ juju consume src:admin/source.juju-qa-test
```

Verification of the save:

```sh
$ juju ssh -m controller 0
$ juju_db_repl
repl (controller)> SELECT * FROM external_controller
uuid                                    alias   ca_cert
ccca535d-fb2a-474e-87f6-e78b107337f5            -----BEGIN CERTIFICATE-----
MIIEEjCCAnqgAwIBAgIUAxmm/LlIBsn6sgHHYyJzXdIpcM8wDQYJKoZIhvcNAQEL
BQAwITENMAsGA1UEChMESnVqdTEQMA4GA1UEAxMHanVqdS1jYTAeFw0yNTA5MTcx
NjA4MjVaFw0zNTA5MTcxNjEzMjVaMCExDTALBgNVBAoTBEp1anUxEDAOBgNVBAMT
B2p1anUtY2EwggGiMA0GCSqGSIb3DQEBAQUAA4IBjwAwggGKAoIBgQDVb8lCeurK
oCZJ9X6cpxNiVCVtQ1yuyyfGfVrGjvzI6vd7fjkUBgMg3Z19H+eRcscJ39jFxZF1
qB3Zgz+X/JK7TGickxiWWxHI6F1esAjVnj2ErMvbA5ozKqTssiW9lpRI36m6qhQX
qhQjuNMFxMJI9rGZNTj5aLHZpK1XyP0808ozI88tJGY3jXwGpUXXBW2Uq0kIaI2f
OF58EeGzcd9VNH/A3G9PI3pTVvxXbbZbtJWLay6L22Ccjyei2NDyshp3g9ctkG5e
934d4OIFGBojWoBzunQqg5DEUfu+Ur5tQ7RJ6IhiutoWXK2IS1GLGWkfOJsnwTGf
h3uqhgGh2ee0YF2R5joNTA2tCLjdnJwOormoerOUK4uITgiUJkT3YAfGbkUb4nRU
fuxsGmXfOO1zo32Zdv+bv6kgIuYJOMF816pXiwsOT731iubJ50VA3uHiyOdO0ISR
3H11NlZM+x7GEQcrX5g7m5rsmaK+ScITnksNWbQC1Cuk7pg6Asl94xMCAwEAAaNC
MEAwDgYDVR0PAQH/BAQDAgKkMA8GA1UdEwEB/wQFMAMBAf8wHQYDVR0OBBYEFCm8
c87C3NnlVORDwHqqHZMpgXFSMA0GCSqGSIb3DQEBCwUAA4IBgQAjkr8bq/05PUKp
INxrqLxlmN5gwWEJ5D9OoAKs0h8rroGUjh2DBPZ9pdPzps8QDNzJFMsI9sODG7/D
pstGidIHpu7RU1+NFnLmNVdVPyJF7ZeXFgt2hlE/EAGTF6XYnNRqEEOlUUiPSVIp
uGG2LjKdZe263sIw7atMXcmQSMT0gIfaIndPP5/J3LD7rMyCdcfRxs6FJPZFccSg
wbKy2/JBu2s7rbGgnAfRsFEfmNI+4Jb9Gx+BrPfTctoEkYBGjy4Sz+Jz7ZnveWnR
sn1g4/syF09C7W5iLKkdulVgmF5iaOiNb1dodWah6fVt2BMISXOzIKHdmjsxsF6A
7SJCRbI402wyMtZ+KaAGOjlbblpLOb8cS9GpaMEvrH9+Vd7zTOU9q9MxOkMSPVBU
9PcGSyCYWHoP3Fr5+LTn40gFS86EDCjXqJotAnwIYuiZrnuy4G3iEZzIFcP+/ZjZ
63WBVjUAbjTjWBMLT7TfCFLNItW38d3MPqBstIY9d5YIHhxs8+4=
-----END CERTIFICATE-----
```

Bonus points, do it on one controller :wink: 

## Links

**Jira card:** [JUJU-6932](https://warthogs.atlassian.net/browse/JUJU-6932)

[JUJU-6932]: https://warthogs.atlassian.net/browse/JUJU-6932?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ

### CODE REVIEW

File: apiserver/facades/client/application/application.go [hunk 1]
```diff
@@ -1336,8 +1344,68 @@ func (api *APIBase) SetRelationsSuspended(ctx context.Context, args params.Relat
...
+	// Maybe save the details of the controller hosting the offer.
+	var externalControllerTag names.ControllerTag
+	if arg.ControllerInfo != nil {
+		externalControllerTag, err = api.saveExternalController(ctx, *arg.ControllerInfo, sourceModelTag)
+		if err != nil {
+			return internalerrors.Errorf("saving external controller info: %w", err)
+		}
+	}
+
+	_ = externalControllerTag
```
Comment (nvinuesa) [inline apiserver/facades/client/application/application.go:1379]: Why? Are you planning to use it later? Otherwise the result from saveExternalController could be ignored right?

**Review by nvinuesa — COMMENTED** [review]
Everything LGTM, except that one bit.

File: apiserver/facades/client/application/application.go (hunk 1 above)
Comment (SimonRichardson) [reply apiserver/facades/client/application/application.go:1379]: I am. It's in a follow up. I didn't want to provide a massive PR of consume done, as it's complicated.

**Review by nvinuesa — APPROVED** [review]

**Review by jack-w-shaw — APPROVED** [review]
